File: Common/Serialization/ObjectTraverser.vb
Namespace Common

    Public Class ObjectTraverser

    End Class

    Public Class ObjectProxy

        Public Sub Reset()
            Me.Container = Nothing
            Me.Type = Nothing
        End Sub

        Public Sub [Set](Of T)(ByVal Value As T)
            Verify.True(Me.Type Is Nothing And Me.Container Is Nothing, "Reset before setting.")
            Me.Type = GetType(T)

            Dim C As ObjectContainer(Of T) = Nothing
            If Me.Containers.TryGetValue(Me.Type, Me.Container) Then
                C = DirectCast(Me.Container, ObjectContainer(Of T))
            Else
                C = New ObjectContainer(Of T)()
                Me.Container = C
                Me.Containers.Add(Me.Type, Me.Container)
            End If

            C.Value = Value
        End Sub

        Public Function [Get](Of T)() As T
            Verify.True(Me.Type IsNot Nothing And Me.Container IsNot Nothing, "Set before getting.")
            Verify.True(Me.Type = GetType(T), "The type set is different.")

            Me.Type = GetType(T)
            Dim C = DirectCast(Me.Container, ObjectContainer(Of T))

            Return C.Value
        End Function

        Private Type As Type
        Private Container As Object
        Private ReadOnly Containers As Dictionary(Of Type, Object) = New Dictionary(Of Type, Object)()

        Private Class ObjectContainer(Of T)

#Region "Value Property"
            Private _Value As T

            Public Property Value As T
                Get
                    Return Me._Value
                End Get
                Set(ByVal Value As T)
                    Me._Value = Value
                End Set
            End Property
#End Region

        End Class

    End Class

End Namespace

File: Common/Serialization/ObjectTraverser.cs
using System.Collections.Generic;
using System;

namespace Ks.Common
{
    public class ObjectTraverser
    {
    }

    public class ObjectProxy
    {
        public void Reset()
        {
            this.Container = null;
            this.Type = null;
        }

        public void Set<T>(T Value)
        {
            Verify.True(this.Type == null & this.Container == null, "Reset before setting.");
            this.Type = typeof(T);

            ObjectContainer<T> C = null;
            if (this.Containers.TryGetValue(this.Type, out this.Container))
            {
                C = (ObjectContainer<T>) this.Container;
            }
            else
            {
                C = new ObjectContainer<T>();
                this.Container = C;
                this.Containers.Add(this.Type, this.Container);
            }

            C.Value = Value;
        }

        public T Get<T>()
        {
            Verify.True(this.Type != null & this.Container != null, "Set before getting.");
            Verify.True(this.Type == typeof(T), "The type set is different.");

            this.Type = typeof(T);
            var C = (ObjectContainer<T>) this.Container;

            return C.Value;
        }

        private Type Type;
        private object Container;
        private readonly Dictionary<Type, object> Containers = new Dictionary<Type, object>();

        private class ObjectContainer<T>
        {
            public T Value { get; set; }
        }
    }
}
